lib/git-process.ts

```diff
@@ -247,3 +251,54 @@ export class GitProcess {
     return null
   }
 }
+
+/**
+ * Prevent errors originating from the stdin stream related
+ * to the child process closing the pipe from bubbling up and
+ * causing an unhandled exception when no error handler is
+ * attached to the input stream.
+ *
+ * The common scenario where this happens is if the consumer
+ * is writing data to the stdin stream of a child process and
+ * the child process for one reason or another decides to either
+ * terminate or simply close its standard input. Imagine this
+ * scenario
+ *
+ *  cat /dev/zero | head -c 1
+ *
+ * The 'head' command would close its standard input (by terminating)
+ * the moment it has read one byte. In the case of Git this could
+ * happen if you for example pass badly formed input to apply-patch.
+ *
+ * Since consumers of dugite using the `exec` api are unable to get
+ * a hold of the stream until after we've written data to it they're
+ * unable to fix it themselves so we'll just go ahead and ignore the
+ * error for them. By supressing the stream error we can pick up on
+ * the real error when the process exits when we parse the exit code
+ * and the standard error.
+ *
+ * See https://github.com/desktop/desktop/pull/4027#issuecomment-366213276
+ */
+function ignoreClosedInputStream(process: ChildProcess) {
+  process.stdin.on('error', err => {
+    const code = (err as ErrorWithCode).code
+
+    // Is the error one that we'd expect from the input stream being
+    // closed, i.e. EPIPE on macOS and EOF on Windows?
+    if (code === 'EPIPE' || code === 'EOF') {
+      return
+    }
+
+    // Nope, this is something else. Are there any other error listeners
+    // attached than us? If not we'll have to mimic the behavior of
+    // EventEmitter.
+    //
+    // See https://nodejs.org/api/errors.html#errors_error_propagation_and_interception
+    //
+    // "For all EventEmitter objects, if an 'error' event handler is not
+    //  provided, the error will be thrown"
+    if (process.stdin.listeners('error').length > 1) {
```

Comment:
> Are there any other error listeners attached than us? If not we'll have to mimic the behavior of EventEmitter.

This mean we should be doing the inverse of this check, right?

```ts
if (process.stdin.listeners('error').length <= 1) {
   throw err
}
```

This feels similar to what `epipebomb` [does](https://github.com/mhart/epipebomb/blob/8c1e6d83a32678f77daf3dad617122291c984517/epipebomb.js#L10-L12):

```js
    // If there's more than one error handler (ie, us),
    // then the error won't be bubbled up anyway
    if (stream.listeners('error').length <= 1) {
      ...
    }
```

Reply:
Doh, right you are. Nice catch.